Triton backward convolution kernel

chien-an-chen opened the pull request:
This PR implements Triton kernels for 2D convolution backward operations (input and weight gradients) in TorchInductor, replacing the previous ATen-only fallback approach.

### Key Changes

  **Core Implementation:**
  - Added Triton templates for `conv2d_bwd_weight` and `conv2d_bwd_input` operations
  - Implemented `convolution_backward_lowering` function with backend selection logic
  - Added layout computation functions for backward operations
  - Added configuration options for backend selection (`max_autotune_conv_bwd_weight_backends`, `max_autotune_conv_bwd_input_backends`)

  **Test Coverage:**
  - Added comprehensive `test_conv2d_backward_parametrized` test covering various parameter combinations (stride, dilation, padding, kernel sizes, groups, NHWC format)
  - Includes proper tolerance handling for ROCM vs other platforms
  - Added skip conditions for unsupported dynamic shapes and subprocess compilation mode
  - Removed `test_conv2d_backward_channels_last_dynamic_shapes` from test failures (now passing)

  ### Performance Impact
  - For certain models, small-batch workloads can achieve up to 20% end-to-end performance improvement on MI325 GPU. Of the 20% performance increase, about 3% comes from reduced kernel execution time, and 17% comes from reduced bubble time.
  - For a single operator, the triton bwd conv kernel has a good chance of outperforming the aten implementation in scenarios input_H=1 input_W=1 kernel_H=1 kernel_W=1 for backward weight, kernel_H=1 kernel_W=1 for backward input. Below are some shape/params that the triton bwd conv kernel has better performance than aten. The data type is bf16, and the tests is run on MI325 GPU.

|type|input shape|grad_out shape|DH|DW|GS|PH|PW|SH|SW|triton_best_time_ms|aten_best_time_ms|
| :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- |
|bwd_weight|3x112x1x1|3x448x1x1|1|1|1|0|0|1|1|0.006200|0.025000|
|bwd_weight|5x224x1x1|5x896x1x1|1|1|1|0|0|1|1|0.006000|0.026400|
|bwd_weight|6x112x1x1|6x448x1x1|1|1|1|0|0|1|1|0.006200|0.027800|
|bwd_weight|2x56x1x1|2x448x1x1|1|1|1|0|0|1|1|0.006700|0.029100|
|bwd_weight|6x896x1x1|6x112x1x1|1|1|1|0|0|1|1|0.006200|0.029600|

|type|grad_out shape|weight shape|DH|DW|GS|PH|PW|SH|SW|triton_best_time_ms|aten_best_time_ms|
| :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- |
|bwd_input|2x224x235x363|224x224x1x1|1|1|1|0|0|1|1|0.061300|0.118300|
|bwd_input|2x448x118x182|448x448x1x1|1|1|1|0|0|1|1|0.057000|0.091100|
|bwd_input|2x224x470x725|224x32x1x1|1|1|1|0|0|1|1|0.088600|0.129100|
|bwd_input|7x224x1x1|224x56x1x1|1|1|1|0|0|1|1|0.006300|0.033800|
|bwd_input|5x224x1x1|224x56x1x1|1|1|1|0|0|1|1|0.007500|0.037900|
|bwd_input|2x112x1x1|112x896x1x1|1|1|1|0|0|1|1|0.006200|0.039000|


cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @jataylo @hongxiayang @naromero77amd @pragupta @jerrymannil @xinyazhang @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168 @aditew01 @gujinghui @PenghuiCheng @jianyuh @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @snadampal @ezyang @EikanWang @wenzhe-nrv @voznesenskym @penguinwu @zhuhaozhe @blzheng @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @Lucaskabela @azahed98 @mlazos @chenyang78

chien-an-chen committed:
[inductor] Add Triton backward convolution with layout optimization

Support lowering backward convolution with Triton kernel.
Includes conv2d_bwd_input_template and conv2d_bwd_weight_template
kernels with autotune support.

chien-an-chen committed:
Fix after rebase on latest main.

chien-an-chen committed:
fix test cases faile in test_torchinductor.py

chien-an-chen committed:
fix lint error. add test_conv2d_backward to test_torchinductor.py

chien-an-chen committed:
fix test fail: DynamicShapesCodegenCpuTests, TestCompiledAutograd

chien-an-chen committed:
fix code style error.

chien-an-chen committed:
fix lint error.

chien-an-chen committed:
fix lint error. add max_autotune flag to test_conv2d_backward.

chien-an-chen committed:
set allow_tf32=False to fix accuracy failure in test.

chien-an-chen committed:
skip some case of test_conv2d_backward_parametrized for dynamice shape.

chien-an-chen committed:
skip test_conv2d_backward_parametrized under subprocess compile mode.

Currently, subprocess mode starts with default TF32 settings
(cudnn.allow_tf32=True) even if parent process has set allow_tf32=False.

eellison reviewed (commented):
@chien-an-chen can you provide more benchmarks on shapes, dtypes, hardware, etc where this is profitable ?

chien-an-chen committed:
Merge remote-tracking branch 'upstream/main' into triton_backward_convolution_kernel

eellison commented on torch/_inductor/kernel/conv.py:
```diff
@@ -685,3 +686,697 @@ def constrain_conv_to_fx_strides(fx_node, *args, **kwargs):
+    )
+
+
+LOOP_BODY_BWD_INPUT_2D = r"""
```
would it make sense to move these to the .jinja template files?

eellison reviewed (commented):
partial review

EricKing626 committed:
Merge branch 'pytorch:main' into triton_backward_convolution_kernel

EricKing626 committed:
Move conv2d backward template from conv.py into Jinja templates

This PR refactors Inductor’s conv2d backward codegen by moving the conv2d_bwd_weight_template  source/LOOP_BODY_BWD_WEIGHT_2D and conv2d_bwd_input_template  source/LOOP_BODY_BWD_INPUT_2D blocks out of torch/_inductor/kernel/conv.py and into the corresponding .py.jinja template files. This keeps the Python lowering code focused on wiring/context and improves readability and maintainability of the conv codegen templates.

EricKing626 committed:
Fix Optional typing and remove unsupported guard_shape in ir_node_to_tensor calls

This commit fixes issues introduced by recent API updates:

Replace deprecated Optional[...] typing with the newer T | None style.

Remove the unsupported guard_shape argument from ir.ir_node_to_tensor() calls.

These changes restore compatibility with the current Inductor APIs and resolve related lint and test failures.

EricKing626 committed:
Add guard_int_seq in torch/_inductor/kernel/conv.py to ensure correct convolution layout guards

This PR makes conv_layout self-contained by using guard_int_seq for convolution parameters, preventing incorrect layouts from symbolic values being silently resolved through size_hints.

EricKing626 committed:
Fix lint error

eellison commented on torch/_inductor/kernel/conv.py:
```diff
@@ -737,3 +738,477 @@ def constrain_conv_to_fx_strides(fx_node, *args, **kwargs):
+        w_shape, dtype=out.dtype, device=x_t.device, memory_format=memory_fmt
+    )
+
+    tmp = torch.ops.aten.convolution_backward(
```
I don't t understand while we're calling convolution backward, and discarding multiple outputs. isn't this going to be wasteful?

eellison commented on torch/_inductor/kernel/conv.py:
```diff
@@ -737,3 +738,477 @@ def constrain_conv_to_fx_strides(fx_node, *args, **kwargs):
+    dw = None
+    choices_dw = []
+    args_w = []
+    layout_dw = conv_bwd_weight_layout(grad_out, input, weight, **kwargs)
```
you could also just do a similar thing as `constrain_conv_bwd_to_fx_strides` - check if we are optimizing to nhwc , or check the output node's stride

eellison reviewed (commented):
a few comments

EricKing626 committed:
Refactored to use process_kernel()

EricKing626 committed:
Update conv.py

EricKing626 committed:
Refactored to use `ExternKernel.process_kernel()` as requested, with Shallow copies

Refactored to use `ExternKernel.process_kernel()` as requested, with workarounds:

- **Shallow copies**: Create temporary `TensorBox` copies to isolate side effects from `realize_input()` and `freeze_layout()`
- **Still heavier**: Adds overhead vs lightweight `ir_node_to_tensor()` + `fake_mode` used in `conv_layout()` (L377)
- **Less clear**: Requires copy creation and unpacking 5-tuple return

**Trade-off**: Consistency with kernel execution path at the cost of complexity and asymmetry with forward conv layout computation.

EricKing626 committed:
Merge branch 'pytorch:main' into triton_backward_convolution_kernel

EricKing626 committed:
Fix lint

jansel reviewed (changes requested):
Fix test failures and address comments from other reviewers

jeffdaily commented:
Closing in favor of #178945, per @EricKing626's note that development moved there to avoid the ROCm/pytorch issue.
